_posts/2021-09-10-essential-dev-experience-concepts-android.md

```diff
@@ -0,0 +1,99 @@
+---
+layout: post
+author: adam
+title: "Five Essential Developer Experience Concepts For Android"
+modified: 2021-09-10
+published: true
+tags: [twitch, toa]
+categories: [android]
+---
+
+In the most recent [live stream](https://www.youtube.com/watch?v=ePpbpLyYI1w) for Tasks Of Affirmation, we looked at creating a good developer experience. It's important to consider developers at the start of any new project, so we can ensure that over the lifetime of a project anyone who contributes can understand how to contribute, match any guidelines that a team has, and get up and running quickly. We also want to ensure that our codebase maintains a certain quality of formatting and a lack of code smells - and putting these checks in place before you even start prevents them from ever being introduced at all. 
+
+Let's look at five developer experience concepts I consider essential to every project.
+
+<!--more-->
+
+During the live stream, we went over the following concepts to use in the TOA application:
+
+* GitHub Actions
+* Danger
+* Ktlint
+* Detekt
+* Git Hooks
+
+The live stream link will have video markers to each of these sections, where you can see the implementation in more detail. For this blog post, we're just going to highlight what each of these concepts are, and why they were necessary for the [TOA application](https://github.com/adammc331/toa).
+
+## GitHub Actions
+
+GitHub Actions is a way to [automate workflows](https://github.com/features/actions) for your GitHub projects. The possibilities are endless - verifying unit tests pass, running lint checks, publishing to the play store, sharing internal builds, generating documentation - any process that you want to automate can be done via GitHub Actions. 
+
+You don't need to use Actions - there are many other CI tools out there such as Bitrise, Circle CI, Jenkins, and many more. We've opted to use GitHub Actions for this project because it is quick and easy to implement straight from GitHub, where this project is hosted anyways. 
+
+### Setup
+
+To create an Action, we need to add a new directory `.github/workflows` in our project. Inside of that, we can add a YAML file that configures the action. This can be edited via the GitHub UI which gives additional tooling for editing them. You can also feel free to use the [TOA Actions](https://github.com/AdamMc331/TOA/tree/development/.github/workflows) for inspiration.
+
+## Danger
+
+Danger is a tool for automating [common pull request chores](https://danger.systems/ruby). There are versions of Danger in multiple languages, but I prefer the Ruby gem. At first this may sound like the same thing as a GitHub Action, but it's a little more specific. Danger will run on Actions (or any other CI tool) to perform checks related to a Pull Request itself. An example of common Danger checks:
+
+* Does this PR have labels?
+* Does this PR have a description?
+* Did we assign a reviewer?
+* Is it too long? 
+* Did we modifile the CHANGELOG.md file?
```

Comment:
Typo with modifile (unless this is a cool new term I don't know!)

Reply:
LMAO